### PR TITLE
storage: mariadb impls

### PR DESCRIPTION
**Pull request by thibault-martinez**

### CODE REVIEW

File: common/storage/sql/mariadb/wrappers.c [hunk 1]
```diff
@@ -6,3 +6,82 @@
...
 
 #include "common/storage/sql/mariadb/wrappers.h"
+
+retcode_t prepare_statement(MYSQL* const db, MYSQL_STMT** const mariadb_statement, char const* const statement) {
+  if (db == NULL || mariadb_statement == NULL || statement == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  if ((*mariadb_statement = mysql_stmt_init(db)) == NULL) {
+    return RC_STORAGE_FAILED_PREPARED_STATEMENT;
```
Comment (tsvisabo) [inline common/storage/sql/mariadb/wrappers.c]: should be different error here

File: common/storage/sql/mariadb/wrappers.c [hunk 2]
```diff
@@ -6,3 +6,82 @@
...
+}
+
+void column_compress_bind(MYSQL_BIND* const bind, size_t const index, void const* const data,
+                          enum enum_field_types const type, size_t const num_bytes) {
+  ssize_t i = num_bytes - 1;
+
+  bind[index].buffer = (void*)data;
+  bind[index].buffer_type = type;
+  if (type == MYSQL_TYPE_BLOB) {
+    for (; i >= 0 && ((flex_trit_t*)data)[i] == FLEX_TRIT_NULL_VALUE; --i) {
```
Comment (tsvisabo) [inline common/storage/sql/mariadb/wrappers.c:82]: TODO?

File: common/storage/sql/mariadb/storage.c [hunk 3]
```diff
@@ -7,13 +7,394 @@
...
+  MYSQL_BIND bind[1];
+
+  memset(bind, 0, sizeof(bind));
+
+  if (mysql_stmt_execute(mariadb_statement) != 0) {
+    log_statement_error(mariadb_statement);
+    return RC_STORAGE_FAILED_EXECUTE;
+  }
+
+  bind[0].buffer = (char*)count;
```
Comment (tsvisabo) [inline common/storage/sql/mariadb/storage.c]: this and the previous has almost the same body, can you share their code?

File: common/storage/sql/mariadb/storage.c [hunk 4]
```diff
@@ -7,13 +7,394 @@
...
+  milestone_reset(milestone);
+
+  bind_out[0].buffer = (char*)&milestone->index;
+  bind_out[0].buffer_type = MYSQL_TYPE_LONGLONG;
+
+  bind_out[1].buffer = (char*)milestone->hash;
+  bind_out[1].buffer_type = MYSQL_TYPE_BLOB;
+  bind_out[1].buffer_length = FLEX_TRIT_SIZE_243;
+
+  if (mysql_stmt_bind_result(mariadb_statement, bind_out) != 0) {
```
Comment (tsvisabo) [inline common/storage/sql/mariadb/storage.c]: can you group both `mysql_stmt_bind_result` `mysql_stmt_store_result` into a *bind_and_store function?

File: common/storage/sql/mariadb/storage.c [hunk 5]
```diff
@@ -7,13 +7,394 @@
...
+  }
+
+  return RC_OK;
+}
+
+static retcode_t storage_transaction_update_generic(MYSQL_STMT* const mariadb_statement, MYSQL_BIND* const bind,
+                                                    flex_trit_t const* const hash) {
+  column_compress_bind(bind, 1, hash, MYSQL_TYPE_BLOB, FLEX_TRIT_SIZE_243);
+
+  if (mysql_stmt_bind_param(mariadb_statement, bind) != 0) {
```
Comment (tsvisabo) [inline common/storage/sql/mariadb/storage.c]: can you group both calls into *bind_and_execute ?

File: common/storage/sql/mariadb/storage.c [hunk 6]
```diff
@@ -31,3 +412,814 @@ retcode_t storage_destroy() {
...
+
+  return execute_statement_count(mariadb_statement, count);
+}
+
+retcode_t storage_transaction_store(storage_connection_t const* const connection,
+                                    iota_transaction_t const* const transaction) {
+  mariadb_tangle_connection_t const* mariadb_connection = (mariadb_tangle_connection_t*)connection->actual;
+  MYSQL_STMT* mariadb_statement = mariadb_connection->statements.transaction_insert;
+  MYSQL_BIND bind[17];
+  uint64_t ts = current_timestamp_ms();
```
Comment (tsvisabo) [inline common/storage/sql/mariadb/storage.c:399]: you can define in the schema file to set to current timestamp instead

File: common/storage/sql/mariadb/storage.c [hunk 7]
```diff
@@ -31,3 +412,814 @@ retcode_t storage_destroy() {
...
+}
+
+retcode_t storage_transaction_update_validity(storage_connection_t const* const connection,
+                                              flex_trit_t const* const hash, bundle_status_t const validity) {
+  mariadb_tangle_connection_t const* mariadb_connection = (mariadb_tangle_connection_t*)connection->actual;
+  MYSQL_STMT* mariadb_statement = mariadb_connection->statements.transaction_update_validity;
+  MYSQL_BIND bind[2];
+
+  memset(bind, 0, sizeof(bind));
+
```
Comment (tsvisabo) [inline common/storage/sql/mariadb/storage.c:552]: can you merge all 3 update functions?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: common/storage/sql/mariadb/storage.c (hunk 7 above)
Comment (thibault-martinez) [reply common/storage/sql/mariadb/storage.c:552]: Not sure what you mean by that, but we can't factorize more due to the type of the parameter

File: common/storage/sql/mariadb/wrappers.c (hunk 2 above)
Comment (thibault-martinez) [reply common/storage/sql/mariadb/wrappers.c:82]: No, the loop is empty just counting the number of ending 9s

File: common/storage/sql/mariadb/storage.c (hunk 6 above)
Comment (thibault-martinez) [reply common/storage/sql/mariadb/storage.c:399]: It was already like that for SQLite3 so I'd need to change the SQL request and sqlite3 storage to make it in the schema

**Review by tsvisabo — DISMISSED** [review]

**Pull request closed without merging**